### ISSUE TITLE
Minio distributed performance on AMD Ryzen 5

### ISSUE THREAD

**Issue by lskbr** [minio/minio#4198]
Hi Guys,

I have a small cluster dedicated to minio. It is composed by 4 machines with 4 x 5T HDD and 1 x 256G SSD. System runs on SSD, minio uses the other four drives.

When I transfer a 1GB file to minio, mc reports 10 MB/s. I'm trying to check if there is performance problem on my setup or another issue.

## Setup

All machines are new and were installed one week ago.

* /etc/default/minio

   MINIO_VOLUMES="http://192.168.1.50/export/1 http://192.168.1.50/export/2 
    http://192.168.1.50/export/3 http://192.168.1.50/export/4 http://192.168.1.51/export/1 
    http://192.168.1.51/export/2 http://192.168.1.51/export/3 http://192.168.1.51/export/4 
    http://192.168.1.52/export/1 http://192.168.1.52/export/2 http://192.168.1.52/export/3 
    http://192.168.1.52/export/4 http://192.168.1.53/export/1 http://192.168.1.53/export/2 
    http://192.168.1.53/export/3 http://192.168.1.53/export/4"
    MINIO_OPTS="--address :9199 --config-dir /etc/minio"

* Version used (`minio version`):
Version: 2017-03-16T21:50:32Z
Release-Tag: RELEASE.2017-03-16T21-50-32Z
Commit-ID: 5311eb22fd681a8cd4a46e2a872d46c2352c64e8

* Linux Distro: Ubuntu 17.04

* Kernel: Linux kitt 4.10.0-20-generic #22-Ubuntu SMP Thu Apr 20 09:22:42 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
* CPU: AMD Ryzen 5 1600 Six-Core Processor
* CPU Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_l2 mwaitx hw_pstate vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 xsaves clzero irperf arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic overflow_recov succor smca
* Ram: 16GB

Disks (for each machine):
* 4 x Toshiba X300 5TB Desktop 3.5 Inch SATA 6Gb/s 7200rpm Internal Hard Drive (https://www.amazon.com/gp/product/B013JPLKQK/ref=oh_aui_detailpage_o02_s00?ie=UTF8&psc=1)

## Performance testes executed:
bigfile is a 1GB file generated from /dev/urandom with dd

* mc cp bigfile minio/test
Performance: 10 MB/s

* scp bigfile minio_server:
Performance: 103 MB/s

* iperf (minio server to another minio server):
```
Client connecting to 192.168.1.50, TCP port 5001
TCP window size: 85.0 KByte (default)
------------------------------------------------------------
[  3] local 192.168.1.51 port 41130 connected with 192.168.1.50 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-10.0 sec  1.08 GBytes   931 Mbits/sec
```

* iperf (another machine in the same network to minio server):
```
Client connecting to 192.168.1.50, TCP port 5001
TCP window size: 85.0 KByte (default)
------------------------------------------------------------
[  3] local 192.168.1.224 port 48744 connected with 192.168.1.50 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-10.0 sec  1.09 GBytes   933 Mbits/sec
```
* Netperf: 927 Mb/s

* Hard disk write performance:
`dd if=/dev/zero of=t.tmp iflag=fullblock,count_bytes count=10G`
200 MB/s

**Comment by fwessels** [comment]
@lskbr Thank you for your report, we are looking into it.

Just a quick question, what is the performance that you get when you copy data out, eg. `mc cp minio/test/bigfile bigfilecopy`?

**Comment by lskbr** [comment]
Hi, I left the cluster running the whole weekend. Performance degraded to unbelievable levels.
Read is at 580KB/s. I have 63GB x 4 in each machine, but even an mc ls is costly.
Processors are at 30~40% usage. All clients were stopped (checked by active connections via netstat), the only active connections are among Minio servers. Local ssh sessions seems normal.

When I try to ls:
X:~$ mc ls Y/Z/data/F
mc: <ERROR> Unable to list folder. You did not provide the number of bytes specified by the Content-Length HTTP header.
Even directories with 3 files take a long time to show.

**Comment by fwessels** [comment]
Thank you for the response, we are investigating and will get back to you.

**Comment by fwessels** [comment]
@lskbr We have seen erratic performance related to caching behaviour (keep objects cached in memory after uploading -- causing the Golang garbage collector to spend a lot of CPU cycles).

Could you try restarting the cluster with caching turned off, e.g.
`_MINIO_CACHE=off minio server ....`

**Comment by deekoder** [comment]
Can we have remote access to your machines @lskbr

**Comment by lskbr** [comment]
Hi! I added _MINIO_CACHE=off
Write performance is now much better: 57MB/s! Read performance still low.
CPU usage is close to 0% now.

Unfortunately no, machines are in a internal VPN.

My test results:
@kitt:~$ mc cp bigtest walle/test/bigtest9
bigtest:       1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 57.69 MB/s 

@kitt:~$ mc cp walle/test/bigtest9 bigtest_w
.../bigtest9:  1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 7.83 MB/s 

@kitt:~$ mc cp walle/test/bigtest9 bigtest_w2
.../bigtest9:  1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 7.84 MB/s 

@kitt:~$ mc cp walle/test/bigtest9 bigtest_w3
.../bigtest9:  1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 7.90 MB/s 

@kitt:~$ mc cp bishop/test/bigtest9 bigtest_w4
.../bigtest9:  1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 14.03 MB/s 

@kitt:~$ mc cp bigtest rosie/test/bigtest10
bigtest:       1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 60.89 MB/s 

@kitt:~$ mc cp rosie/test/bigtest10 bigtest_w5
...bigtest10:  1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 9.83 MB/s 

@kitt:~$ mc cp bishop/test/bigtest10 bigtest_w5
...bigtest10:  1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 9.80 MB/s

**Comment by fwessels** [comment]
Thanks, these results are more like it, please keep the cluster running and let's see if the problem does not reappear (CPU usage stays low).

The read performance is still poor. We noticed that you are running an AMD Ryzen CPU, it could be that it might be related to bit-rot hashing that we have optimized using assembly -- could you run some benchmarking for the BLAKE2B algorithm? Please see https://github.com/minio/blake2b-simd, in this directory you can run the benchmarks like this (provided you have Golang installed):
```
$ go test -bench=.
BenchmarkComparisonMD5-8       	    1000	   1579941 ns/op	 663.68 MB/s
BenchmarkComparisonSHA1-8      	    1000	   2118937 ns/op	 494.86 MB/s
BenchmarkComparisonSHA256-8    	     300	   5483614 ns/op	 191.22 MB/s
BenchmarkComparisonSHA512-8    	     500	   3281594 ns/op	 319.53 MB/s
BenchmarkComparisonBlake2B-8   	    1000	   1493910 ns/op	 701.90 MB/s
BenchmarkSize64-8              	 3000000	       402 ns/op	 158.83 MB/s
BenchmarkSize128-8             	 5000000	       336 ns/op	 379.96 MB/s
BenchmarkSize1K-8              	 1000000	      1627 ns/op	 629.27 MB/s
BenchmarkSize8K-8              	  100000	     11243 ns/op	 728.62 MB/s
BenchmarkSize32K-8             	   30000	     42982 ns/op	 762.35 MB/s
BenchmarkSize128K-8            	   10000	    175051 ns/op	 748.76 MB/s
PASS
```

**Comment by lskbr** [comment]
Here are the results (go 1.6):
```
@kipp:~/blake2b-simd$ go test -bench=.
PASS
BenchmarkComparisonMD5-8    	    1000	   1323031 ns/op	 792.56 MB/s
BenchmarkComparisonSHA1-8   	    1000	   2237161 ns/op	 468.71 MB/s
BenchmarkComparisonSHA256-8 	     300	   5474363 ns/op	 191.54 MB/s
BenchmarkComparisonSHA512-8 	     500	   3777828 ns/op	 277.56 MB/s
BenchmarkComparisonBlake2B-8	    1000	   1835128 ns/op	 571.39 MB/s
BenchmarkSize64-8           	 2000000	       590 ns/op	 108.45 MB/s
BenchmarkSize128-8          	 3000000	       512 ns/op	 249.62 MB/s
BenchmarkSize1K-8           	 1000000	      2090 ns/op	 489.73 MB/s
BenchmarkSize8K-8           	  100000	     13987 ns/op	 585.66 MB/s
BenchmarkSize32K-8          	   30000	     56388 ns/op	 581.11 MB/s
BenchmarkSize128K-8         	   10000	    217474 ns/op	 602.70 MB/s
ok  	/blake2b-simd	22.393s
```

**Comment by lskbr** [comment]
I kept the cluster running. After sometime with 100 clients, I got 25G ram and an unstoppable minio. I could not stop it. Once I killed minio with killall it restarted with normal CPU and memory usage.

Regarding the benchmark:
```
@walle:~/blake2b-simd$ go test -bench=.
BenchmarkComparisonMD5-12        	    1000	   1508736 ns/op	 695.00 MB/s
BenchmarkComparisonSHA1-12       	    1000	   1222719 ns/op	 857.58 MB/s
BenchmarkComparisonSHA256-12     	     500	   2705683 ns/op	 387.55 MB/s
BenchmarkComparisonSHA512-12     	     500	   2796766 ns/op	 374.92 MB/s
BenchmarkComparisonBlake2B-12    	    1000	   1469141 ns/op	 713.73 MB/s
BenchmarkSize64-12               	 3000000	       438 ns/op	 146.02 MB/s
BenchmarkSize128-12              	 5000000	       394 ns/op	 324.40 MB/s
BenchmarkSize1K-12               	 1000000	      1670 ns/op	 613.03 MB/s
BenchmarkSize8K-12               	  200000	     11223 ns/op	 729.89 MB/s
BenchmarkSize32K-12              	   30000	     44262 ns/op	 740.31 MB/s
BenchmarkSize128K-12             	   10000	    174885 ns/op	 749.48 MB/s
PASS
ok  	/blake2b-simd	20.588s
```

**Comment by lskbr** [comment]
I make new tests with minio built from source:
Version: 2017-05-01T19:14:51Z
Release-Tag: DEVELOPMENT.2017-05-01T19-14-51Z
Commit-ID: 6cf6828a4c28ed43c67028da6a0bb60f9e002a8e

Write performance is better. Cache still disabled. Read performance not ok.
I noticed that mc takes a while before starting to transfer, at least a few seconds.

```
@kitt:~$ mc cp bigtest walle/test/big_13
bigtest:       1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 59.82 MB/s 

@kitt:~$ mc cp bigtest walle/test/big_14
bigtest:       1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 62.05 MB/s 

@kitt:~$ mc cp walle/test/big_13 b_13w
...st/big_13:  1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 9.80 MB/s 

@kitt:~$ mc cp bishop/test/big_13 b_13w
...st/big_13:  1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 9.67 MB/s 

@kitt:~$ mc cp rosie/test/big_13 b_13w
...st/big_13:  1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 9.60 MB/s
```
From the same machine:
```
@kitt:~$ mc cp kitt/test/big_13 b_13w
...st/big_13:  1024.00 MB / 1024.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 14.90 MB/s
```
I have everything set to recompile and install minio. If you need more tests, just let me know.

**Comment by harshavardhana** [comment]
This looks like either erasure decoding or blake2b is causing some problems on your hardware.

**Comment by fwessels** [comment]
As reported by @lskbr we are seeing low performance on this AMD hardware for certain benchmarks (especially for reconstructing):
```
$ go test -run=NONE -bench=.
BenchmarkEncode10x2x10000-12                   50000         36067 ns/op    2772.57 MB/s
BenchmarkEncode100x20x10000-12                  2000        814700 ns/op    1227.45 MB/s
BenchmarkEncode17x3x1M-12                       1000       1310789 ns/op    13599.27 MB/s
BenchmarkEncode10x4x16M-12                        30      43462485 ns/op    3860.16 MB/s
BenchmarkEncode5x2x1M-12                        5000        318599 ns/op    16456.04 MB/s
BenchmarkEncode10x2x1M-12                       3000        775004 ns/op    13529.93 MB/s
BenchmarkEncode10x4x1M-12                       2000       1108932 ns/op    9455.73 MB/s
BenchmarkEncode50x20x1M-12                        50      27604667 ns/op    1899.27 MB/s
BenchmarkEncode17x3x16M-12                        50      39420041 ns/op    7235.22 MB/s
BenchmarkVerify10x2x10000-12                   30000         59475 ns/op    1681.36 MB/s
BenchmarkVerify50x5x50000-12                    2000        762830 ns/op    6554.54 MB/s
BenchmarkVerify10x2x1M-12                       2000       1132731 ns/op    9257.05 MB/s
BenchmarkVerify5x2x1M-12                        2000        830774 ns/op    6310.84 MB/s
BenchmarkVerify10x4x1M-12                       1000       1621910 ns/op    6465.07 MB/s
BenchmarkVerify50x20x1M-12                       100      14891369 ns/op    3520.75 MB/s
BenchmarkVerify10x4x16M-12                        30      45816725 ns/op    3661.81 MB/s
BenchmarkReconstruct10x2x10000-12              20000         90585 ns/op    1103.93 MB/s
BenchmarkReconstruct50x5x50000-12               2000       1456863 ns/op    3432.03 MB/s
BenchmarkReconstruct10x2x1M-12                  1000       1504130 ns/op    6971.31 MB/s
BenchmarkReconstruct5x2x1M-12                   2000       1088586 ns/op    4816.23 MB/s
BenchmarkReconstruct10x4x1M-12                  1000       2315367 ns/op    4528.77 MB/s
BenchmarkReconstruct50x20x1M-12                   50      22752101 ns/op    2304.35 MB/s
BenchmarkReconstruct10x4x16M-12                   20      59482383 ns/op    2820.54 MB/s
BenchmarkReconstructP10x2x10000-12             50000         27438 ns/op    3644.46 MB/s
BenchmarkReconstructP50x5x50000-12                 1    1631713430 ns/op       3.06 MB/s
BenchmarkReconstructP10x2x1M-12                    1    4044891325 ns/op       2.59 MB/s
BenchmarkReconstructP5x2x1M-12                     1    2076468269 ns/op       2.52 MB/s
BenchmarkReconstructP10x4x1M-12                    1    4091109094 ns/op       2.56 MB/s
BenchmarkReconstructP50x20x1M-12                   1    20677747273 ns/op       2.54 MB/s
BenchmarkReconstructP10x4x16M-12                   1    65888316383 ns/op       2.55 MB/s
BenchmarkStreamEncode10x2x10000-12               200       9320101 ns/op      10.73 MB/s
BenchmarkStreamEncode100x20x10000-12              20      74633545 ns/op      13.40 MB/s
BenchmarkStreamEncode17x3x1M-12                  100      18384150 ns/op     969.63 MB/s
BenchmarkStreamEncode10x4x16M-12                  20      55872489 ns/op    3002.77 MB/s
BenchmarkStreamEncode5x2x1M-12                   200       7030709 ns/op     745.71 MB/s
BenchmarkStreamEncode10x2x1M-12                  100      11217779 ns/op     934.74 MB/s
BenchmarkStreamEncode10x4x1M-12                  100      13433599 ns/op     780.56 MB/s
BenchmarkStreamEncode50x20x1M-12                  20      79640782 ns/op     658.32 MB/s
BenchmarkStreamEncode17x3x16M-12                  20      86251245 ns/op    3306.77 MB/s
BenchmarkStreamVerify10x2x10000-12               200       9846583 ns/op      10.16 MB/s
BenchmarkStreamVerify50x5x50000-12                30      33802782 ns/op     147.92 MB/s
BenchmarkStreamVerify10x2x1M-12                  100      10960992 ns/op     956.64 MB/s
BenchmarkStreamVerify5x2x1M-12                   200       7033301 ns/op     745.44 MB/s
BenchmarkStreamVerify10x4x1M-12                  100      12522453 ns/op     837.36 MB/s
BenchmarkStreamVerify50x20x1M-12                  30      52814480 ns/op     992.70 MB/s
BenchmarkStreamVerify10x4x16M-12                 100      20457316 ns/op    8201.08 MB/s
PASS
ok      /reedsolomon    191.562s
```

**Comment by fwessels** [comment]
@lskbr Just for completeness, can you also pls check golang.org/x/crypto/blake2b, eg.:

```
$ pwd
/Users/frankw/golang/src/golang.org/x/crypto/blake2b
$ go test -bench=.
BenchmarkWrite128-8   	10000000	       225 ns/op	 567.66 MB/s
BenchmarkWrite1K-8    	 1000000	      1576 ns/op	 649.45 MB/s
BenchmarkSum128-8     	 5000000	       244 ns/op	 523.17 MB/s
BenchmarkSum1K-8      	 1000000	      1541 ns/op	 664.25 MB/s
```

**Comment by fwessels** [comment]
This error looks related to the AMD Ryzen in some way. We will check if we can get access to a server with such an AMD CPU to investigate further.

**Comment by deekoder** [comment]
@fwessels is there any update to this issue?

**Comment by fwessels** [comment]
See above, we need to decide if we want to get access to a server with an AMD Ryzen CPU.

**Comment by fwessels** [comment]
Based on profile logs that we obtained this looks to be a low-level go runtime channel closing/synchronization issue that is specifically happening on this type of AMD Ryzen CPU. We have not seen this on other CPUs.

The customer already mentioned that access to the servers is difficult due to internal VPN issues (@lskbr any updates?), and due to the difficulty of getting access to a server with an AMD Ryzen CPU, we will close the issue for now.

We will reopen the issue if we can get access to an AMD Ryzen or the same issue is reported in similar situations.

**Comment by deekoder** [comment]
Closing issue due to lack of access to this particular CPU. @lskbr please reopen if you can help us with access.
